refactor: refact the pouch hook plugin (#2360)
alibaba/pouch: docs/features/pouch_with_plugin.md

> # Pouch-container with plugin

Reviewer: Why to change the name of it? 
I think the logo of this project is PouchContainer indeed.

Author: Should pouchcontainer instead of pouch-container?

	// could change the container create body passed-in by http request body
	PreCreate(*types.ContainerCreateConfig) error

	// PreStart returns an array of priority and args which will pass to runc, the every priority
	// used to sort the pre start array that pass to runc, network plugin hook always has priority value 0.
	PreStart(interface{}) ([]int, [][]string, error)

	// PreCreateEndpoint accepts the container id and env of this container and returns the priority of this endpoint
	// and if this endpoint should enable resolver and a map which will be used as generic params to create endpoints of
	// this container
	PreCreateEndpoint(string, []string) (priority int, disableResolver bool, genericParam map[string]interface{})

	// PreUpdate defines plugin point where receives a container update request, in this plugin point user
	// could change the container update body passed-in by http request body
	PreUpdate(io.ReadCloser) (io.ReadCloser, error)

	// PostUpdate called after update method successful,
	// the method accepts the rootfs path and envs of container
	PostUpdate(string, []string) error

> } []()

Reviewer: remove `[]()`?

## Which plugins

Currently pouch provides four plugins，they are: `container plugin`，`daemon plugin`，`volume plugin`，`cri plugin`

### container plugin

In order to run custom code provided by users which will be triggered at some point, we support a plugin framework.

* pre-create container point, at this point you can change the input stream by some rules, in some company they have some stale orchestration system who use env to pass-in some limit which is an attribute in PouchContainer, then you can use this point to convert value in env to attribute in ContainerConfig or HostConfig of PouchContainer create api.
* pre-start container point, at this point you can set more pre-start hooks to oci spec, where you can do some special thing before container entrypoint start, priority decide the order of executing of the hook. libnetwork hook has priority 0, so if the hook is expected to run before network in container setup you should set priority to a value big then 0, and vice versa.
* pre-create-endpoint container point, at this point you can return the priority of this endpoint and if this endpoint need enable resolver and the generic params of this endpoint.

> * pre-update container point, at this point you can change the input stream as you want format or rules.

Reviewer 2: ```suggestion
* pre-update container point, at this point you can change the input stream to your wanted format or rules.
```

	// could change the container create body passed-in by http request body
	PreCreate(*types.ContainerCreateConfig) error

	// PreStart returns an array of priority and args which will pass to runc, the every priority
	// used to sort the pre start array that pass to runc, network plugin hook always has priority value 0.
	PreStart(interface{}) ([]int, [][]string, error)

	// PreCreateEndpoint accepts the container id and env of this container and returns the priority of this endpoint
	// and if this endpoint should enable resolver and a map which will be used as generic params to create endpoints of
	// this container
	PreCreateEndpoint(string, []string) (priority int, disableResolver bool, genericParam map[string]interface{})

	// PreUpdate defines plugin point where receives a container update request, in this plugin point user
	// could change the container update body passed-in by http request body
	PreUpdate(io.ReadCloser) (io.ReadCloser, error)

	// PostUpdate called after update method successful,
	// the method accepts the rootfs path and envs of container
	PostUpdate(string, []string) error
}
```

### daemon plugin

> * pre-start daemon point, at this point you can start assist processes like network plugins and dfget proxy which need by pouchd and whose life cycle is the same as pouchd.

Reviewer 2: ```suggestion
* pre-start daemon point, at this point you can start some assistant processes like network plugins and dfget proxy which are needed by pouchd and whose life cycle is the same as pouchd.
```

> * pre-stop daemon point, at this point you can stop the assist processes gracefully, but the trigger of this point is not a promise, because pouchd may be killed by SIGKILL.

Reviewer 2: ```suggestion
* pre-stop daemon point, at this point you can stop the assistant processes gracefully, but the trigger of this point is not a promise, because pouchd may be killed by SIGKILL.
```

Defined as follow:

```
// DaemonPlugin defines places where a plugin will be triggered in pouchd lifecycle
type DaemonPlugin interface {
	// PreStartHook is invoked by pouch daemon before real start, in this hook user could start dfget proxy or other
	// standalone process plugins

> 	PreStartHook() error

Reviewer 2: What about just simplify this name to `PreStart` `PreStop` ?

	// PreStopHook is invoked by pouch daemon before daemon process exit, not a promise if daemon is killed, in this
	// hook user could stop the process or plugin started by PreStartHook
	PreStopHook() error
}
```

### volume plugin

> * pre-volume-create volume point, at this point you can change the create volume's config as you want, add your default volume's options.

Reviewer 2: ```suggestion
* pre-volume-create volume point, at this point you can change the volume's create config as you want, add your default volume's options.
```

Defined as follow:

```
// VolumePlugin defines places where a plugin will be triggered in volume lifecycle
type VolumePlugin interface {
	// PreVolumeCreate defines plugin point where receives an volume create request, in this plugin point user
	// could change the volume create body passed-in by http request body
	PreVolumeCreate(*types.VolumeCreateConfig) error
}
```

### cri plugin

> * pre-create-container cri container point, at this point you can update the container config what it will be created, such as update the container's envs or labels.

Reviewer 2: ```suggestion
* pre-create-container cri point, at this point you can update the container config to what it will be created, such as update the container's envs or labels.
```

	// PreCreateContainer defines plugin point where receives a container create request, in this plugin point user
	// could the container's config in cri interface.
	PreCreateContainer(*types.ContainerCreateConfig, interface{}) error
}
```

## Example

### How to write

Introduce how to write a daemon plugin.

#### 1. Make your plugin package

You can make your package in directory `hookplugins/daemonplugin` and add a go file `daemon_hook.go`.

#### 2. Define a struct for your plugin

In `daemon_hook.go` define your plugin object, struct is `type daemonPlugin struct{}`.

#### 3. Register your plugin

In `init` function to register your plugin, now we provide 4 plugin to register:

    * `RegisterContainerPlugin`
    * `RegisterDaemonPlugin`
    * `RegisterCriPlugin`
    * `RegisterVolumePlugin`

In my plugin, we use `RegisterDaemonPlugin` to register a daemon plugin into pouch daemon.

```
func init() {
	hookplugins.RegisterDaemonPlugin(&daemonPlugin{})
}
```

#### 4. Implement the plugin's interface function

We implement the daemon plugin's interface function, we just print some logs.

```
// PreStartHook copy config from /etc/pouch/config.json to /etc/sysconfig/pouch

> // and start plugin processes than daemon depended

Reviewer 2: mismatched doc.

func (d *daemonPlugin) PreStartHook() error {
	logrus.Infof("Call my daemon plugin pre start hook")
	return nil
}

> // PreStopHook stops plugin processes than start ed by PreStartHook.

Reviewer 2: ```suggestion
// PreStopHook stops plugin processes than started by PreStartHook.
```
